### ISSUE TITLE
Out of memory for object context when doing simple RSA key operations in Python

### ISSUE THREAD

**Issue by bnevis-i** [tpm2-software/tpm2-pkcs11#567]
## Environment

Board: APL UPSquared with PTT
OS: Ubuntu 20.04 (opensc 0.20.0-3)
tpm2-tss: 4.2.1 (tarball)
tpm2-pkcs11: 1.3.1 (tarball)
tpm2-tools: 4.1.3 (tarball)
tpm2-abrmd: Not installed
Token initialized with tpm2_ptool with label tpmhsm

RSA keypair generated with
`tpm2_ptool addkey --id  (someid) --userpin (redacted) --algorithm rsa2048 --label tpmhsm --key-label rsakey

## Script
```
import os
import pkcs11

lib = pkcs11.lib('libtpm2_pkcs11.so')

token = lib.get_token(token_label='tpmhsm')

with token.open(rw=False, user_pin='123456') as session:
    print(session)

    public = session.get_key(label='rsakey',object_class=pkcs11.constants.ObjectClass.PUBLIC_KEY)
    print(public)

    private = session.get_key(label='rsakey',object_class=pkcs11.constants.ObjectClass.PRIVATE_KEY)
    print(private)

    plaintext = '1234567890123456'

    ciphertext = public.encrypt(plaintext, mechanism=pkcs11.mechanisms.Mechanism.RSA_PKCS)
    print(ciphertext)

    plaintext = 'notdecrypted'
    plaintext = private.decrypt(ciphertext, mechanism=pkcs11.mechanisms.Mechanism.RSA_PKCS)
    print(plaintext)
```

## Error message
```
WARNING:esys:src/tss2-esys/api/Esys_TestParms.c:269:Esys_TestParms_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_TestParms.c:95:Esys_TestParms() Esys Finish ErrorCode (0x000001c4) 
WARNING:esys:src/tss2-esys/api/Esys_TestParms.c:269:Esys_TestParms_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_TestParms.c:95:Esys_TestParms() Esys Finish ErrorCode (0x000001c4) 
WARNING:esys:src/tss2-esys/api/Esys_TestParms.c:269:Esys_TestParms_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_TestParms.c:95:Esys_TestParms() Esys Finish ErrorCode (0x000001e6) 
WARNING:esys:src/tss2-esys/api/Esys_TestParms.c:269:Esys_TestParms_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_TestParms.c:95:Esys_TestParms() Esys Finish ErrorCode (0x000001e6) 
WARNING:esys:src/tss2-esys/api/Esys_TestParms.c:269:Esys_TestParms_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_TestParms.c:95:Esys_TestParms() Esys Finish ErrorCode (0x000001e6) 
WARNING:esys:src/tss2-esys/api/Esys_TestParms.c:269:Esys_TestParms_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_TestParms.c:95:Esys_TestParms() Esys Finish ErrorCode (0x000001e6) 
<pkcs11._pkcs11.Session object at 0x7f4d0a919e50>
<PublicKey label='rsakey' id='6565386332373336303538333433353539303335666136343031393236643839' 2048-bit RSA>
<PrivateKey label='rsakey' id='6565386332373336303538333433353539303335666136343031393236643839' 2048-bit RSA>
b'\x0c:\xb3-O\xd9X\xc2\x97JFX\xef\\\xaaPjt\xbe\xec\x8e\x88\x07\xc1\x11\xd4\xd4z\x1fm\xef\x1a\xf2z :y\x96\x1e\xc1\xcd\xc6\x1b*\xd0n;y>%\xd7^E\x17\xb8\xa4\xcb\xcb\xa0($\xf3\x95\x8b}\x9c\xad\xf0A`_\xf3\xa9\xff\x02\x86](n"9%\xe3\xa0}\x8f2\x9f\x06\x06\xe5I\x03\xf5\x94/\xbay\xe7\xa9\xe5J\x89>\x9bZ\x17Q\xbc\x94\x10\nj\x8f\xfbr~\xbb\xc2\xf0\x80g\xcd\xaf\xd8\x18\xa68\xde\xb8T4\xac\xb7\x8b_\xca\xa4\xe8=4M\x1f&-\xfe\xackdf\xbd-c#b\xa9O\xbb\x18L\xf8\x8b\x06\xa3\x85\xb6\xd8\x9c7z\x96\x1d`\x05\x96I]Ii\xe8B\xd5\xb3\x92\xba\xdf/p\xac\xa9\xd9\x0e\xa6C_\x10\xff\xdbe[\xd5z\x1a\x99\x95\xa5\x0c\xd28i\x0f\xe7\xb9\xcdr\x15i\x0bs\xf2\x8e\xfaJ;\xca\xc3\xcb\xfb\x02mF\xc3j\xb9\xd2\xf4\x1cZ%\xd6L\x8a?|]\xc7\x1d\t\xa16h\x9d\x93\xac8n'
WARNING:esys:src/tss2-esys/api/Esys_Load.c:320:Esys_Load_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_Load.c:112:Esys_Load() Esys Finish ErrorCode (0x00000902) 
ERROR: Esys_Load: tpm:warn(2.0): out of memory for object contexts:
WARNING:esys:src/tss2-esys/api/Esys_FlushContext.c:234:Esys_FlushContext_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_FlushContext.c:89:Esys_FlushContext() Esys Finish ErrorCode (0x000001c4) 
ERROR: Esys_FlushContext: tpm:parameter(1):value is out of range or is not correct for the context
Traceback (most recent call last):
  File "tpm_rsa.py", line 23, in <module>
    plaintext = private.decrypt(ciphertext, mechanism=pkcs11.mechanisms.Mechanism.RSA_PKCS)
  File "/home/(redacted)/.local/lib/python3.8/site-packages/pkcs11/types.py", line 899, in decrypt
    return self._decrypt(data, **kwargs)
  File "pkcs11/_pkcs11.pyx", line 845, in pkcs11._pkcs11.DecryptMixin._decrypt
  File "pkcs11/_pkcs11.pyx", line 846, in pkcs11._pkcs11.DecryptMixin._decrypt
  File "pkcs11/_errors.pyx", line 88, in pkcs11._pkcs11.assertRV
pkcs11.exceptions.GeneralError
```

**Comment by williamcroberts** [comment]
Are you using a resource manager? What's your TCTI configuration?

**Comment by bnevis-i** [comment]
The test configuration is not overriding the TCTI and the TPM2 resource broker is not installed.  I have validated through strace that it is defaulting to the kerne resource manager (/dev/tpmrm0)

**Comment by bnevis-i** [comment]
@williamcroberts no issues when running with tpm2-abrmd.

**Comment by williamcroberts** [comment]
> @williamcroberts no issues when running with tpm2-abrmd.
> 
> Kernel: 5.4.0-42-generic #46-Ubuntu

That screams of a bug in the in-kernel RM. We would want to search through the mailing archives and if we
cannot find anything related or determine the fix, send a mail to:
https://kernsec.org/wiki/index.php/Linux_Kernel_Integrity

Make sure we CC Jarrko, but he is out of office at the moment returning in a couple of weeks.

**Comment by tstruk** [comment]
Problem is it is not a bug but a current design of the in kernel RM. It's very limited functionality comparing to the user space RM.

**Comment by bnevis-i** [comment]
I consider this a usability issue as a USB HSM does not require a resource broker to function.  It would be nice if there were some manual resource management that could be done that could be done at the application level to make the resource broker daemon optional.

**Comment by williamcroberts** [comment]
> Problem is it is not a bug but a current design of the in kernel RM. It's very limited functionality comparing to the user space RM.

Huh? We're not exercising any abrmd specific extensions in tpm2-pkcs11. It's literally load transient objects and use them. If they didn't implement management of transient objects, flush and re-load for the client, the kernel doesn't have much of a resource manager at all.

**Comment by williamcroberts** [comment]
> I consider this a usability issue as a USB HSM does not require a resource broker to function. It would be nice if there were 

Not really familiar with it's internals but perhaps it or something else is doing the RM functionality and you're not seeing it?

> some manual resource management that could be done that could be done at the application level to make the resource broker daemon optional.

Their is no requirement to make a resource manager a service, but that's generally how it's been done. You could bake it into the tpm2-tss code itself, the application, etc. I chose *not* to implement it in the application, because that's the most complicated place to add it.

An easy way to do it would be with a TCTI that peers into the command stream and does the mapping the RM does.

**Comment by williamcroberts** [comment]
@bnevis-i I started a thread with the kernel-integrity mailing list, they want to know if you have any logs from the kernel around this, anything?

**Comment by bnevis-i** [comment]
@williamcroberts I will try to reproduce it (been through a couple of patch cycles since the original report).

**Comment by bnevis-i** [comment]
@williamcroberts No luck.  Console is clean.  dmesg is clean.  Passed "debug" flag to kernel and ran "dmesg -n 8" as well after a reboot before running the test case.

**Comment by williamcroberts** [comment]
Darn, I'll have to figure out a way to reproduce and get a bug report to the kernel side.

**Comment by williamcroberts** [comment]
@bnevis-i can you share the init steps for "Token initialized with tpm2_ptool with label tpmhsm" so I can replicate the exact environment. Just dummy out any sensitive parts like pins.

**Comment by bnevis-i** [comment]
Bad news.  Can no longer reproduce but now getting a different error.

Also of possible interest as that I found a stray tpm2_pkcs11.sqlite3 file in my working directory with the python script.  I had remove the sqlite3 file in my home dir and in the python script dir (and created the new token in my home dir).

``
./tpm2_ptool.py init
./tpm2_ptool.py addtoken --pid=1 --sopin=(redacted) --userpin=(redacted) --label=tpmhsm ./tpm2_ptool.py addkey --id `uuidgen` --userpin (redacted) --algorithm rsa2048 --label tpmhsm --key-label rsakey
python3 tpm_rsa.py  (my script above)

<pkcs11._pkcs11.Session object at 0x7fe34a6bf940>
<PublicKey label='rsakey' id='66666164623266392d643533332d343133372d626566342d613634636430363135376137' 2048-bit RSA>
<PrivateKey label='rsakey' id='66666164623266392d643533332d343133372d626566342d613634636430363135376137' 2048-bit RSA>
Traceback (most recent call last):
  File "tpm_rsa.py", line 19, in <module>
    ciphertext = public.encrypt(plaintext, mechanism=pkcs11.mechanisms.Mechanism.RSA_PKCS)
  File "/home/bnevis/.local/lib/python3.8/site-packages/pkcs11/types.py", line 867, in encrypt
    return self._encrypt(data, **kwargs)
  File "pkcs11/_pkcs11.pyx", line 770, in pkcs11._pkcs11.EncryptMixin._encrypt
  File "pkcs11/_pkcs11.pyx", line 775, in pkcs11._pkcs11.EncryptMixin._encrypt
  File "pkcs11/_errors.pyx", line 88, in pkcs11._pkcs11.assertRV
MemoryError: Buffer was too small. Should never see this.
```

**Comment by williamcroberts** [comment]
> Bad news. Can no longer reproduce but now getting a different error.

Hrmmm.....

> Also of possible interest as that I found a stray tpm2_pkcs11.sqlite3 file in my working directory with the python script. I had  remove the sqlite3 file in my home dir and in the python script dir (and created the new token in my home dir).

You need to set TPM2_PKCS11_PATH or it will default to different things..... per https://github.com/tpm2-software/tpm2-pkcs11/blob/master/docs/INITIALIZING.md

If no existing store is found, it will:
1. If env variable TPM2_PKCS11_STORE is set, attempt to use that path directory or create it if it doesn't exist.
   On failure, it continues to number 2.
2. /etc/tpm2_pkcs11 or whatever was configured at build time with --with-storedir.
3. if $HOME is set, attempts to use that path directory. If the directory doesn't exist it will be created.
   This almost always exceeds for most users, so this ends up as the default store most of the time. If it fails,
   continues on to number 4.
4. Use the Current Working Directory.

**Comment by williamcroberts** [comment]
Let me see if I can reproduce that error, your tpm2-tss version seems suspect as 4.2.1 doesn't exist, did you mean 2.4.1?

**Comment by bnevis-i** [comment]
@williamcroberts Yes, 2.4.1.  And the only in $HOME is the only sqlite3 DB that is hanging around.

**Comment by williamcroberts** [comment]
I can say with v1.3.1 of tpm2-pkcs11, I am seeing issues in the DB location logic I described above. It seems that it's using CWD instead of $HOME location.... It's probably best to always use TPM2_PKCS11_STORE env var and --path

I got to the point of the script running, and mine just hangs trying to encrypt with the public key:
```python
ciphertext = public.encrypt(plaintext, mechanism=pkcs11.mechanisms.Mechanism.RSA_PKCS)
```
I don't see any output, it just hangs....

However, I updated to 1.4.0 and have replicated your error, will start diving deeper into it.

**Comment by williamcroberts** [comment]
INFO on line: "499" in file: "src/pkcs11.c": return "C_Encrypt" value: 336

So the caller is passing in a return buffer of size 16.... I think all other callers have always used keysize..... which in this case **will** work, let me try fixing that.

**Comment by williamcroberts** [comment]
On version 1.4.0, this patch fixes the buffer too small issue:
```diff
diff --git a/src/lib/ssl_util.c b/src/lib/ssl_util.c
index 82c01d7..f499785 100644
--- a/src/lib/ssl_util.c
+++ b/src/lib/ssl_util.c
@@ -300,11 +300,6 @@ CK_RV ssl_util_encrypt(EVP_PKEY *pkey,
         return CKR_GENERAL_ERROR;
     }
 
-    if ((CK_ULONG)to_len > *ctextlen) {
-        *ctextlen = to_len;
-        return CKR_BUFFER_TOO_SMALL;
-    }
-
     EVP_PKEY_CTX *pkey_ctx = EVP_PKEY_CTX_new(pkey, NULL);
     if (!pkey_ctx) {
         LOGE("OOM");
```

However, it seems to be hanging on the:
```python
plaintext = private.decrypt(ciphertext, mechanism=pkcs11.mechanisms.Mechanism.RSA_PKCS)
```

Ill have to keep digging into that....

**Comment by williamcroberts** [comment]
I can detect output buffer too small by openssl:
```C
 #include <openssl/err.h>
 #include <openssl/evperr.h>

int rc = EVP_PKEY_encrypt(ctx, out, &outlen, (const unsigned char *)in, inlen);
if (rc <= 0) {
	 unsigned long r = ERR_get_error();
	 printf("FAIL: %lu - %s\n", r, ERR_error_string(r, NULL));
	 int reason = ERR_GET_REASON(r);
	 printf("reason: %d\n", reason);
	 if (reason == EVP_R_BUFFER_TOO_SMALL) {
		 printf("Yep CKR_BUFFER_TOO_SMALL\n");
		 return 1;
	 }
}
```

I can just set outlet to keylen in that case...

**Comment by williamcroberts** [comment]
So what I did was set up a working environment and non-working environment. They both have the same versions of:
- tpm2-tss: 3.0.1
 - tpm2-tools: 4.1.3
 - tpm2-pkcs11: master

**Working Machine**:
- ubuntu 18.04
- Python 3.8.1
- ibmtpm1628
- tpm2-arbmd: 2.3.3-rc0

***Broken Machine***:
- Ubuntu 20.04
- Python 3.8.5
- qemu emulated, using StefanBerger SWTPM interface.

The latest BUG is that buffer too small error. I commented that check out, but it's really the proper check. OpenSSL refuses to perform that operation with the output buffer size set at 16 bytes.... looking at the test code for this python module, I don't even see them testing this case, public encrypt with pkcs11.mechanisms.Mechanism.RSA_PKCS:
  - https://github.com/danni/python-pkcs11/blob/master/tests/test_rsa.py

The buffer size has to be keylength size... so not sure what to do there.

What I have noticed between the machines is that the code leaves the Python interface and jumps into the binding. It leaves the python interface with the same arguments to self._encrypt:
```python
        if isinstance(data, bytes):
            print("self._encrypt: len(data){} kwargs: {}".format(len(data), kwargs))
            return self._encrypt(data, **kwargs)
```
You see it in the output as:
`self._encrypt: len(data)16 kwargs: {'mechanism': <Mechanism.RSA_PKCS>}`
On both machines.

I patched the pkcs11 module to print some more info as well. However, what we see is different data being used to call into the pkcs11 module.

On the broken machine we see:
`BILL: CTEXTLEN: 16`

On the working machine we see:
`BILL: CTEXTLEN: 139846147882768`

This makes me believe their is something wonky going on in the bindings. To confirm this, I mirrored my "working" setup on the "broken" machine by setting up the ibmtpm simulator and tpm2-abrmd. Since this was a qemu image, I was able to fully drop the /dev/tpm0 and /dev/tpmrm0 devices to ensure that they are not being used. I saw the same errors.

**My best guess is this is a binding issue between cython versions.**

The initial working case, ctextlen seems ridiculously large: 139846147882768. It smells like an uninitialized variable. Then in the broken version goes to 16, which seems ok, but is too small, as padding RSA_PKCS requires at a minimum the key size in bytes - 11, per OpenSSL:
```
 to must point to a memory section large enough to hold the maximal possible decrypted data (which is equal to RSA_size(rsa) for RSA_NO_PADDING, RSA_size(rsa) - 11 for the PKCS #1 v1.5 based padding modes and RSA_size(rsa) - 42 for RSA_PKCS1_OAEP_PADDING). padding is the padding mode that was used to encrypt the data. to and from may overlap.
```
To further this along, I tried changing the plaintext to something keysize... as well as change the mechanism to raw RSA and it worked:
```python
with token.open(rw=False, user_pin='123456') as session:
    print(session)

    public = session.get_key(label='rsakey',object_class=pkcs11.constants.ObjectClass.PUBLIC_KEY)
    print(public)

    private = session.get_key(label='rsakey',object_class=pkcs11.constants.ObjectClass.PRIVATE_KEY)
    print(private)

    plaintext = 'QEUvEQbem8OvEE27Pw0WOF0dsUaglmGpInaJ1AZilqm5VzpBjvCYtCgGmqV91AC3JpP4GP3YTtCsnKMOG9xZlz3jO7oqVHCAsmnSdL8g80FWV1N8yRHHWSYoEQ4ropWdpQCABt82xGCIKwwVggQM0VTJgHOlpNDCBUdfwPor7MsN1QAmQk53vHZfqfEHWSPe92UV8WlTloLl2sbmGPRiSFP16BjuyJreZIoBGZOPOvKMKjI77GxuHPV0bRVpoOCF'

    ciphertext = public.encrypt(plaintext, mechanism=pkcs11.mechanisms.Mechanism.RSA_X_509)
    print(ciphertext)

    plaintext = 'notdecrypted'
    plaintext = private.decrypt(ciphertext, mechanism=pkcs11.mechanisms.Mechanism.RSA_X_509)
    print(plaintext)
```

I took that working example and ran it with the tpmrm0 node in my qemu environment, and that worked as well.

It might be good to bring in that maintainer and see if he has suggestions. I also don't see tests for this cipher mode RSA_PKCS. We should probably add tests or file a bug against that project.

### TL;DR
1. plaintext set to 256 bytes and mechanism RSA_X_509 works for /dev/tpmrm0 and tpm2-abrmd environments
2. their seems to be differences in arguments to C_<Encrypt|Decrypt> functions likely from differences in generated cython wrapper code between Ubuntu 18.04 and Ubuntu 20.04 machines.
3. RSA_PKCS11 mechanism appears untested in python-pkcs11 package
4. I cannot reproduce the out of memory errors on Ubuntu 20.04 using /dev/tpmrm0

**Comment by bnevis-i** [comment]
I am content to close this as I can no longer reproduce the original issue and the later source of the errors appear to be the Python package itself.

**Comment by bnevis-i** [comment]
@williamcroberts Fabulous work on the debugging.

**Comment by williamcroberts** [comment]
> @williamcroberts Fabulous work on the debugging.

My head hurts.I can say as a work-around, I just use ctypes to inetrface with pkcs11. It's not as nice, but it can be a stop gap till we get this sorted. Here is a gist with some code to get you started if you choose to go that route....
  - https://gist.github.com/williamcroberts/c996ad1d53979115a281118f8c3630ec

@bnevis-i are you going to contact the maintainer of that other package or shall I? I don't want to close this until someone else has the ball firmly. I have had to answer these questions for this other module a few times already, hence why I have that gist.

**Comment by bnevis-i** [comment]
> @bnevis-i are you going to contact the maintainer of that other package or shall I? I don't want to close this until someone else has the ball firmly. I have had to answer these questions for this other module a few times already, hence why I have that gist.

I've moved on to other activities and don't see myself coming back to this in the near-term future.  I would be appreciative if you raised the escalation.

**Comment by williamcroberts** [comment]
@bnevis-i will do

**Comment by williamcroberts** [comment]
Oh it is being tested BTW:
`test_sessions.py:73:            crypttext = pub.encrypt(data, mechanism=pkcs11.Mechanism.RSA_PKCS)`

**Comment by williamcroberts** [comment]
@bnevis-i  errrr.... I figured it out with fresh eyes and while writing up the bug report for the other project.

The python bindings call C_Encrypt once to get the buffer size, and once to actually do the encryption. When I added in public key operations that get performed with OSSL, I broke that, "call with null buffer and ill give you the size" convention. So I was performing the encryption, incorrectly.

This fixes it for me, want to give it a spin?
```diff
diff --git a/src/lib/ssl_util.c b/src/lib/ssl_util.c
index 9971cb773f84..3aa4659b98f2 100644
--- a/src/lib/ssl_util.c
+++ b/src/lib/ssl_util.c
@@ -294,6 +294,11 @@ CK_RV ssl_util_encrypt(EVP_PKEY *pkey,
 
     CK_RV rv = CKR_GENERAL_ERROR;
 
+    if (!ctext) {
+        *ctextlen = EVP_PKEY_size(pkey);
+        return CKR_OK;
+    }
+
     EVP_PKEY_CTX *pkey_ctx = EVP_PKEY_CTX_new(pkey, NULL);
     if (!pkey_ctx) {
         LOGE("OOM");
```

I am going to get some tests around this today and submit a PR hopefully soon.

**Comment by bnevis-i** [comment]
Validated against 1.4.0 release as having fixed the issue.
